crates/bevy_ecs/src/query/builder.rs

```diff
@@ -422,6 +423,89 @@ mod tests {
         }
     }
 
+    #[test]
+    fn builder_provide_access() {
+        let mut world = World::new();
+        world.spawn((A(0), B(1)));
+
+        let mut query =
+            QueryBuilder::<(Entity, FilteredEntityRef, FilteredEntityMut)>::new(&mut world)
+                .data::<&mut A>()
+                .data::<&B>()
+                .build();
+
+        // The `FilteredEntityRef` only has read access, so the `FilteredEntityMut` can have read access without conflicts
```

Comment:
The access is provided from left to right, so if the tuple had been (FilteredEntityMut,  FilteredEntityRef), the Ref wouldn't have read access to the mutable component?

It is not super obvious that the order matters, maybe this should be documented slightly more in the FilteredEntityRef/Mut docstrings?

Reply:
> The access is provided from left to right, so if the tuple had been (FilteredEntityMut, FilteredEntityRef), the Ref wouldn't have read access to the mutable component?

Yes, that's right!  

> It is not super obvious that the order matters, maybe this should be documented slightly more in the FilteredEntityRef/Mut docstrings?

I wanted to ensure that it was *sound* to use more than one `FilteredEntity` thing in a query, but it's not really *useful*, and I don't expect anyone to actually want to do it.  So I'd be inclined to leave the exact semantics out of the documentation so as not to confuse normal users with the details of unimportant edge cases.

Reply:
That's reasonable